### PR TITLE
Refactor net_core/connection inbound packet handling

### PR DESCRIPTION
**Pull request by ghost**
While working on other pull requests, I noticed that Zephyr's current net_core/connection inbound packet pipeline is rather hard to read, understand and change.

It is also strongly coupled to IP protocols although there have been alternative protocols introduced to the code base (e.g. custom IEEE 802.15.4 and CAN) and more are easily implemented in userspace based on AF_PACKET. This breaks protocol encapsulation and unnecessarily bloats code size/flash size for custom non-IP protocols.

That's why I propose some iterative refactoring.

My goal is to make the code more readable and gradually decouple core network code from IP without breaking anything in userspace (i.e. without API changes - except for one small change which has already been reviewed and was not considered harmful by the code maintainers).

### CODE REVIEW

File: subsys/net/ip/connection.c
Comment (rlubos) [inline subsys/net/ip/connection.c:564]: While not a strict requirement, we tend to declare local variables at the beginning of the block, I wouldn't change that, unless necessary.

File: subsys/net/ip/connection.c
Comment (rlubos) [inline subsys/net/ip/connection.c:612]: It's not clear to me, why would this differentiation be needed? If `CONFIG_NET_SOCKETS_PACKET_DGRAM` is not enabled,  `net_packet_socket_input()` won't be called after the L2 was processed (with ETH_P_ALL proto).

File: subsys/net/ip/connection.c
Comment (rlubos) [inline subsys/net/ip/connection.c:602]: It'd be good to add comments to closing `#endif`, which `#if` do they correspond to, it's easy to get lost if they're nested (`/* defined(CONFIG_NET_IP) */` in this case for example).

File: include/zephyr/net/net_context.h
Comment (rlubos) [inline include/zephyr/net/net_context.h:607]: Theoretically, we should deprecate the old function name, and make it call the new renamed function, I think however in this particular case it's ok to just rename, `net_context()` isn't really intended to be used by applications so I guess we could consider it internal.

File: subsys/net/ip/net_context.c
Comment (rlubos) [inline subsys/net/ip/net_context.c:273]: I understand that using `IS_ENABLED()` with switch/case statements is problematic, but here? Why changing it, the compiler should rule out that code anyway?

**Review by rlubos — COMMENTED** [review]
I personally find the packet matching logic within `net_conn_input()` one of the most complex and hard-to-follow thing in our networking stack, an initial impression from my side regarding this PR is that this is step in a good direction. I will need however thorough testing once it's in shape to do so, since this is one of the essential features of the network stack.

**Comment by rlubos** [comment]
@jukkar I know you're probably busy, but it'd be great if you could also have a look at this proposal.

**Comment by fgrandel** [comment]
@rlubos I'm very grateful that you took the considerable time to provide me with a preliminary review and even specific feedback which will of course be integrated as soon as possible, probably this weekend. And yes - I totally agree that thorough testing (and documentation of the underlying architectural direction) is required before a merge. I just wanted to be sure that the general approach finds acceptance with the community before I invest all the time to make this change watertight.

File: include/zephyr/net/net_pkt.h
Comment (fgrandel) [reply include/zephyr/net/net_pkt.h:201]: Fixed.

File: subsys/net/ip/Kconfig
Comment (fgrandel) [reply subsys/net/ip/Kconfig:11]: Fixed.

File: subsys/net/ip/Kconfig
Comment (fgrandel) [reply subsys/net/ip/Kconfig:22]: Fixed.

File: subsys/net/ip/net_core.c
Comment (fgrandel) [reply subsys/net/ip/net_core.c:125]: I added AF_UNSPEC to this switch case - but there are other places where I introduced switches based on address family - some of which do not have the same problem, others might. I'll therefore leave this comment unresolved. This needs to be tested thoroughly before a merge.

File: subsys/net/ip/connection.c
Comment (fgrandel) [reply subsys/net/ip/connection.c:564]: Yes you're right - reverted. This was just a leftover of me trying to identify opportunities to improve code encapsulation (e.g. introducing smaller sub-functions with less local variables).

File: subsys/net/ip/connection.c
Comment (fgrandel) [reply subsys/net/ip/connection.c:612]: Good observation - no longer needed. Removed the variable and further simplified code.

File: subsys/net/ip/connection.c
Comment (fgrandel) [reply subsys/net/ip/connection.c:602]: Fixed here and in all other places where I introduced the variable and where a nested #if exists.

File: include/zephyr/net/net_context.h
Comment (fgrandel) [reply include/zephyr/net/net_context.h:607]: Ok, I renamed the method then without deprecating it and mentioned the decision in the commit comment so it can be reverted if it causes trouble.

File: subsys/net/ip/net_context.c
Comment (fgrandel) [reply subsys/net/ip/net_context.c:273]: It's certainly a question of taste, therefore I reverted the change but my personal arguments are:
1) IMHO It's more obvious what is intended if you use simple preprocessor code without clever non-standard constructs unless such constructs provide an advantage that could not otherwise be achieved.
2) Code that is intended to always be removed should IMHO be eliminated by the preprocessor not by the optimizer. This prevents unintended side effects (e.g. flash/memory waste) when optimization is voluntarily disabled (e.g. to improve debugging experience).
3) Only #if / #endif has full IDE support. Most IDEs will dim code that is eliminated by the preprocessor, improve code navigation, etc. These advantages are lost when using non-standard constructs.

Just to let you know what I was thinking and to answer your question, not to start any argument of course!

File: subsys/net/ip/net_core.c
Comment (fgrandel) [reply subsys/net/ip/net_core.c:126]: @rlubos After doing quite some tests I believe that this problem should be solved.

**Comment by fgrandel** [comment]
@rlubos @nandojve @jukkar This branch is now stable and tested on real hardware. I believe it is ready for final review.

I wouldn't merge this just before a release but if it goes into the mainline in the beginning of a release cycle it can get the necessary coverage by the community, I guess.

File: subsys/net/ip/net_pkt.c
Comment (jukkar) [inline subsys/net/ip/net_pkt.c:1772]: Instead of adding these `#ifdef` here and there, we could add them to `net_pkt_set_ip_hdr_len()` 
so something like
```
void net_pkt_set_ip_hdr_len(pkt, len)
{
#if defined(CONFIG_NET_IP)
set len here....
#endif
}
```

File: subsys/net/ip/CMakeLists.txt
Comment (jukkar) [inline subsys/net/ip/CMakeLists.txt:39]: Typo in commit "net: lib: sockets: improve scalability through **addtl** hidden var" subject.
Perhaps we could say here `net: lib: sockets: improve scalability through a hidden var` instead

File: subsys/net/ip/canbus_socket.c
Comment (jukkar) [inline subsys/net/ip/canbus_socket.c:23]: Hmm, why this change? Asserts can be turned off so we do not then have the family check.

File: subsys/net/ip/connection.c
Comment (jukkar) [inline subsys/net/ip/connection.c:508]: We have been trying to put code info blocks that can be compile checked i.e., so using the `IS_ENABLED()` macro when possible. We should be able to still use that macro here like in the original code.

File: subsys/net/ip/connection.c
Comment (jukkar) [inline subsys/net/ip/connection.c:518]: Similar case here. If we add the `IS_ENABLED(CONFIG_NET_SOCKETS_PACKET)` where this function `conn_raw_socket()` is called, then this ifdef check can be removed and this function gets compile tested most of the time.

**Review by jukkar — COMMENTED** [review]
Looks reasonable. One thing I noticed is that you removed lot of those `IS_ENABLED()` calls which we have been trying to add so that the compiler would test all the sections for us and we could avoid bit rotting the code that is rarely enabled. I would rather see that we do not remove these macros.
I did not fully review the changes to connection.c logic as it is quite complex and gives me headache by just looking into the code ;) but any changes there are most welcome if it makes the code more readable.

**Review by rlubos — COMMENTED** [review]
Overall looks good, thanks for addressing my previous comments. Good to see all of the automated tests pass, I've also verified a few samples manually, all seem to work fine. Not approving yet though until remaining open discussions are resolved.

File: subsys/net/ip/connection.c
Comment (fgrandel) [reply subsys/net/ip/connection.c:508]: @jukkar First of all, thank you so much for taking the time to review this stuff. And yeah, I understand why you prefer IS_ENABLED() - to be honest I didn't fully understand the rationale behind it until now. I did this to prepare for factoring this stuff out to separate files in the future. But of course this can (and should) be done then. There is a trade off to IS_ENABLED() but taking into account what you're saying I understand what the advantages are and will revert as many of these instances as possible. So: good point!

Comment (fgrandel) [reply subsys/net/ip/connection.c]: Fixed everywhere.

File: subsys/net/ip/connection.c
Comment (fgrandel) [reply subsys/net/ip/connection.c:511]: Fixed here as well as in two similar cases in the same file.

File: subsys/net/ip/net_pkt.c
Comment (fgrandel) [reply subsys/net/ip/net_pkt.c:1772]: Fixed here as well as for the corresponding getter.

File: subsys/net/ip/CMakeLists.txt
Comment (fgrandel) [reply subsys/net/ip/CMakeLists.txt:39]: Fixed.

File: subsys/net/ip/canbus_socket.c
Comment (fgrandel) [reply subsys/net/ip/canbus_socket.c:23]: As this function is not exposed in userspace it will only be called by the kernel itself. But then my change introduces family checks BEFORE calling family-specific socket APIs as it never makes sense to call `net_*some-family*_input()` methods for packets that do not belong to the corresponding family.

In other words: This check has not been removed but moved further upstream to the single (single) call site. I also do not see any reason why this method should ever be called elsewhere (and if so the assertion should surface the problem - which would be a kernel bug).

Hope this makes sense?

**Comment by fgrandel** [comment]
@rlubos @nandojve @jukkar I believe that I've addressed and resolved all review comments. Can you have another look and - if you agree - merge this stuff? As this is really a complex change I'll of course stick around to support it in case any negative side effects, bugs or edge cases will be discovered later on.

**Review by jukkar — APPROVED** [review]
Thanks! LGTM.

**Review by rlubos — APPROVED** [review]
Thank you, LGTM

**Pull request merged**
